### ISSUE TITLE
FromDevice element - VLAN tag stripped out.

### ISSUE THREAD

**Issue by TuanThai** [kohler/click#295]
Hello Click folks,

I run Click in user-level, it seems that FromDevice element cannot receive entire of ethernet frames, i.e.,  VLAN tags are stripped out. Are there any ways to receive the frames with VLAN tag?

Thanks a lot for your help.
Tuan

**Comment by bcronje** [comment]
There must be something wrong with your config, FromDevice should and does receive the entire ethernet frame.

Can you post FromDevice config you are running?

Also, check that the interface you are using does not have vlan offloading enabled on the interface, e.g. `ethtool -k eth0`, it is best to disable any ip/tcp/vlan offloading on the NIC.

**Comment by ahenning** [comment]
And also enable "modprobe 8021q"

**Comment by TuanThai** [comment]
I just create a simple configuration file:

`FromDevice(h2-eth0, SNIFFER false, PROMISC true) -> ToDump(test.dump) -> Queue -> ToDevice(h2-eth0);`

I already turn off offloading functions on the interface and enable "8021q". However, there is no vlan header in packets captured in the "test.dump" file. I also use Wireshark to capture network traffic, and it does see vlan headers in incoming packets.

Thanks!

**Comment by ahenning** [comment]
It could be ToDump. Replace ToDump with Print.

**Comment by TuanThai** [comment]
It's the same. I add ToDump just for debuging. I think vlan tab stripped out by nic driver. Are there any ways to disable vlan stripping from nic driver?

Thanks a lot!

**Comment by ahenning** [comment]
If it was on the NIC driver level I wonder how wireshark got the tags.  Just shooting off ideas:

Try setting pcap explicitly:
`wan0 :: FromDevice(wan0, SNIFFER false, PROMISC true, METHOD PCAP);`

Try configuring with --enable-etherswitch

We use the VLANEncap elements and so far I think all the options we set to support VLANs are been listed in this thread.

**Comment by ahenning** [comment]
Ever found the cause of the missing VLAN tags? I would be interested to know which scenario, cards or drivers to avoid.

**Comment by TuanThai** [comment]
After carefully testing, I believe that VLAN tag is stripped out by NIC driver before Click or Wireshark gets it. But Wireshark has support to get it back. This explains why FromDevice element cannot see VLAN tags. Some Windows drivers support getting VLAN tags, but not Linux drivers.

https://wiki.wireshark.org/CaptureSetup/VLAN
http://www.intel.com/content/www/us/en/support/network-and-i-o/ethernet-products/000005498.html

**Comment by ahenning** [comment]
Interesting, thanks. What card and driver is this?

**Comment by tbarbette** [comment]
I also saw that most cards strip it and put the VLAN number in some metadata, but you should be able to disable that? I don't know in which annotation it would be in userlevel-pcap. I wanted to add that for DPDK I/O elements, as it is quite easy. For Netmap, it would require modifying Netmap... VLAN encapsulation offloading would be interesting too.

**Comment by ahenning** [comment]
VLAN tags appear to be relatively reliable on userlevel using pcap. At least with the few common Intel and Realtek 1gbe cards I tested. The exception is with virtual machines where the hypervisor needs to support passing the tags along.

**Comment by tbarbette** [comment]
I may be wrong but there is no code copying the VLAN ID from the pcap header to the VLAN annotation currently in Click right? I don't even know how to retrieve it using PCAP... This would need to be added.

**Comment by ahenning** [comment]
The VLANDecap and SetVLANAnno elements can set the Click VLAN annotation.

**Comment by tbarbette** [comment]
Yes but if the VLAN header has been stripped out by the NIC, it will not work. This needs proper support in the input element.

**Comment by ahenning** [comment]
> If you are capturing on the host system where the VLANs are configured, you will probably not see the VLAN tags in the captured frames -- even if you capture on the physical device. The driver is stripping the tags before the pcap library sees them

Possibly why I cannot replicate the stripped VLAN scenario.

**Comment by pallas** [comment]
When using click to intercept traffic, I always just disable 802.1Q support in the kernel. YMMV

**Comment by justuswilhelm** [comment]
I've been encountering a similar issue with tap devices using KernelTap:

1. Click sends 802.1ad packets to `KernelTap(DEVNAME tap_name)` device.
2. Second click process picks up packets using FromDevice with `METHOD` set to `LINUX` as click was not compiled with pcap headers available.
3. Packets are read in on `FromDevice(tap_name)`, but connecting it to `ToDump` shows the VLAN tags clearly missing.
4. Wireshark connected to the same `tap_name` clearly shows the VLAN tags being present.
5. Recompiling click with libpcap-dev installed on Debian solved the problem.

Now in hindsight it might be obvious that PCAP does a better job at preserving VLAN tags, but getting there took me the better part of a day. So: Developer beware when using FromDevice and make sure click is compiled with libpcap. Disabling or enabling the 8021q kernel module has not made a difference here.

**Comment by tbarbette** [comment]
Well it's probably just that somehow using PCAP disable the VLAN tag removing by the NIC. But you can also disable it with ethtool and Click will see them, without using PCAP.

Anyway using PCAP is better because it uses newer socket facilities that are faster than the one in click (AF_PACKET; Packet_mmap, ...)

**Comment by justuswilhelm** [comment]
@tbarbette Thanks for your suggestion and help. What I did before is to run `ethtool -K local txvlan off`, but it didn't give me the result I desired. Perhaps it is necessary to make sure to use ethtool before using FromDevice. Unfortunately I was running out of time to be able to try it out. Is there another invocation that might work? Also, good to know that there are real performance benefits in using PCAP 👍

**Comment by larsbro** [comment]
You can specify the METHOD like this:
FromDevice(mytap, PROMISC true, METHOD PCAP) ->
ToDump(/tmp/mytap_pcap.pcap) ->
Discard

// Apparently removes the outer tag
FromDevice(mytap, PROMISC true, METHOD LINUX) ->
ToDump(/tmp/mytap_linux.pcap) ->
Discard

**Comment by ahenning** [comment]
without modprobe 8021q you might see the ifconfig dropped packets counter increasing
